handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java

```diff
@@ -608,70 +650,73 @@ public synchronized SSLEngineResult unwrap(
         }
 
         // Write encrypted data to network BIO
-        int bytesConsumed = -1;
-        try {
-            while (srcsOffset < srcsEndOffset) {
+        int bytesConsumed = 0;
+        if (srcsOffset < srcsEndOffset) {
+            do {
                 ByteBuffer src = srcs[srcsOffset];
                 int remaining = src.remaining();
                 int written = writeEncryptedData(src);
-                if (written >= 0) {
-                    if (bytesConsumed == -1) {
-                        bytesConsumed = written;
-                    } else {
-                        bytesConsumed += written;
-                    }
+                if (written > 0) {
+                    bytesConsumed += written;
+
                     if (written == remaining) {
                         srcsOffset ++;
-                    } else if (written == 0) {
-                        break;
                     }
                 } else {
+                    // BIO_write returned a negative or zero number, this means we could not complete the write
+                    // operation and should retry later.
+                    // TODO: Should we use BIO_should_retry() to get the exact cause ?
```

Comment:
Lets update this comment and instead indicate why we do not use `BIO_should_retry`.  And also consider adding logging incase our assumptions are wrong.